WcaOnRails/app/controllers/api/v0/competitions_controller.rb

```diff
@@ -1,11 +1,85 @@
 # frozen_string_literal: true
 class Api::V0::CompetitionsController < Api::V0::ApiController
-  def show
-    competition = Competition.where(id: params[:id], showAtAll: true).first
+  def competition_from_params
+    id = params[:competition_id] || params[:id]
+    competition = Competition.find_by_id(id)
+
+    # If this competition exists, but is not publicly visible, then only show it
+    # to the user if they are able to manage the competition.
+    if competition && !competition.showAtAll && !can_manage_competition?(competition)
+      competition = nil
+    end
+
     unless competition
-      render json: { error: "Competition with id #{params[:id]} not found" }, status: 404
-      return
+      render json: { error: "Competition with id #{id} not found" }, status: 404
+      return nil
+    end
+    competition
+  end
+
+  def raise_unless_manage_competitions_scope
```

Comment:
Yeah, the exception pattern looks nice! I didn't think of it, but this _is_ what exceptions are for.